### ISSUE TITLE
Remove the `.count` fallback that uses find

### ISSUE THREAD

**Issue by kevinburkeshyp** [balderdashy/waterline#1142]
If an adapter does not define a `.count` method, Waterline [will attempt to find() all records and then measure the length of the resulting array][count].

In the best possible case, this is slow and inefficient. In the worst case, you're pulling so many records into memory that you're causing the node to swap. This is particularly pernicious because it may be fine when you first deploy and only hit you after 3 months, when the table has grown and you're wondering why your memory usage is growing more each day.

If the adapter doesn't specify a `.count` method Waterline should just throw an error. 

[count]: https://github.com/balderdashy/waterline/blob/master/lib/waterline/adapter/dql.js#L201

**Comment by devinivy** [comment]
Waterline implements several polyfills to offer a common interface among various databases, some of which are inherently memory- or cpu-intensive in worst case scenarios.

One option for improvement would be for waterline to select on the primary key if possible, so that the entire record may not have to be fetched into memory.  It's also worth pointing-out that this strategy is no more memory-intensive than a `find` with `criteria`, and I imagine that feature will remain in waterline.

My personal take is that we should document this better in the waterline docs and encourage adapter-authors to tell their users that `count` is not implemented.  If an adapter author wants, they are also free to block use of `count` by implementing a method that throws.

This is up for community discussion!  Would love to hear @tjwebb @dmarcelino @particlebanana's thoughts here!  Thanks for bringing this up @kevinburkeshyp.

**Comment by kevinburkeshyp** [comment]
> It's also worth pointing-out that this strategy is no more memory-intensive than a find with criteria, and I imagine that feature will remain in waterline.

When I do a find with criteria, I expect lots of records to be returned and that the query will consume lots of memory. When I do a `.count` I assume the ORM is using the database's built in `count()` method, and am unpleasantly surprised when it does not do so.

**Comment by devinivy** [comment]
I'm sure a PR to the relevant adapter would be appreciated and would at least solve the immediate problem at hand.  Waterline naturally can't infer the best method to count records for any given database on its own, and all the officially supported adapters for databases that have native counting functionality do implement `count`.

**Comment by tjwebb** [comment]
> If the adapter doesn't specify a .count method Waterline should just throw an error.

If it's possible to count something, even if it's via `.find()`, it should do so. I think it'd be more appropriate to throw a warning so that the developer can then make an informed decision about whether they would like to continue using the `.find()` fallback, or if they would like to improve the underlying adapter.

**Comment by particlebanana** [comment]
I'm down with removing the `count` polyfill. It's actually one that can't be polyfilled very good.

I also agree that we need a better documented/explained adapter spec. There was a repo you could fork and start filling in the methods but it's really outdated.

**Comment by devinivy** [comment]
@particlebanana sails-generate-adapter is around and more up to date than sails-adapter-boilerplate.  If you want to do this, should we schedule it for 0.11?  sails-memory and sails-disk will need updating since they rely on the polyfill.

**Comment by particlebanana** [comment]
It would have to be in `0.11` because of the breaking changes.

**Comment by devinivy** [comment]
I meant 0.11 versus WL_Next / future releases.

**Comment by particlebanana** [comment]
So what's the consensus on this? Remove the polyfill or keep it? If we remove it there will be an error returned for adapters that don't support it.

**Comment by devinivy** [comment]
Could we move this to a collection option similar to the one that allows/disallows creation of dynamic the `findByAttributeName()` methods?

**Comment by dmarcelino** [comment]
I'm with @tjwebb on this:

> If it's possible to count something, even if it's via .find(), it should do so. I think it'd be more appropriate to throw a warning [...]

For someone developing an adapter it means less API tests breaking from the get go. For someone using an adapter that doesn't implement count it means he has the basic functionality even if it's in the best performing way.

**Comment by kevinburkeshyp** [comment]
> they have the basic functionality even if it's not* in the best performing way.

The worry here is that anything you do will be performant when you are just getting started and there are no records in the datastore.. it's only 4 months or a year in when your memory usage starts spiking, either crashing your process or making it really slow, and you get curious why, since you didn't implement any changes that seem like they could have triggered it. 

The ORM is not an obvious place to look.

I'm not sure that moving it to a collection option is much better. Either you have `.count()` queries in your code or you don't, and if you do, you're going to have memory spikes in a year or 2 years. You can toggle the setting at that point, but you're still going to need to rewrite the code to not use the `.count()` functionality.

**Comment by devinivy** [comment]
I see what you're saying.  Is the issue in your particular case that the data store chosen doesn't have an efficient way to count records, or that the adapter didn't implement a performant solution?

**Comment by kevinburkeshyp** [comment]
> Is the issue in your particular case that the data store chosen doesn't have an efficient way to count records, or that the adapter didn't implement a performant solution?

We're using Postgres, which has a COUNT() function that returns one value instead of all rows. It [can be slow in some cases][slow], but is strictly faster than fetching all rows and iterating over them, because the DB can avoid all of the read() calls, and only needs to send one value over the wire (among other optimizations)

[slow]: https://wiki.postgresql.org/wiki/Slow_Counting

**Comment by sailsbot** [comment]
Thanks for posting, @kevinburkeshyp.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
We didn't come to a conclusion here.  If we need to address this, we should decide on something and make a PR to the roadmap if necessary.
